### PR TITLE
Add surface reconstruction filter

### LINKED ISSUE #1614: vtkSurfaceReconstructionFilter: reconstruct surface from point cloud around 3D volume
Over time, I have seen a ton of questions asking how to triangulate the surface of a 3D object from a point cloud. As many examples have demonstrated, this is tough to do with delaunay triangualation:

- https://github.com/pyvista/pyvista-support/issues/397
- https://github.com/pyvista/pyvista-support/issues/213
- https://github.com/pyvista/pyvista-support/issues/170
- https://github.com/pyvista/pyvista-support/issues/178
- and many more...

This task is actually handled quite well by the [`vtkSurfaceReconstructionFilter` filter in VTK](https://vtk.org/doc/nightly/html/classvtkSurfaceReconstructionFilter.html#details) and we should work to make this a new filter in PyVista for `PolyData` meshes. Here is example usage:

```py
import pyvista as pv
import numpy as np
from pyvista import _vtk as vtk

points = pv.wrap(pv.Sphere().points)

alg = vtk.vtkSurfaceReconstructionFilter()
alg.SetInputDataObject(points)
alg.Update()
mc = vtk.vtkMarchingCubes()
mc.SetInputDataObject(alg.GetOutput())
mc.SetValue(0, 0.0)
mc.Update()
surf = pv.wrap(mc.GetOutput())
surf.clear_data()

p = pv.Plotter(shape=(1,2))
p.add_mesh(points)
p.add_title('Point Cloud of 3D Surface')
p.subplot(0,1)
p.add_mesh(surf, color=True, show_edges=True)
p.add_title('Reconstructed Surface')
p.show()
```

Screen Shot 2021-08-28 at 1 28 00 PM

### PR DESCRIPTION
**Pull request by akaszynski**
Resolves #1614 by adding a surface reconstruction filter.  Nearly identical implementation as in #1614.

### CODE REVIEW

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -4417,3 +4417,82 @@ def reflect(self, normal, point=None, inplace=False,
...
+        -------
+        pyvista.PolyData
+            Reconstructed surface.
+
+        Examples
+        --------
+        Create a point cloud out of a sphere and reconstruct a surface
+        from it.
+
+        >>> import pyvista
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:4466]:
```suggestion
        >>> import pyvista as pv
```

File: examples/01-filter/surface_reconstruction.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+###############################################################################
+# Plot the point cloud and the reconstructed sphere.
+
+pl = pv.Plotter(shape=(1,2))
+pl.add_mesh(points)
+pl.add_title('Point Cloud of 3D Surface')
+pl.subplot(0,1)
+pl.add_mesh(surf, color=True, show_edges=True)
+pl.add_title('Reconstructed Surface')
+pl.show()  # doctest:+SKIP
```
Comment (tkoyama010) [inline examples/01-filter/surface_reconstruction.py:30]:
```suggestion
pl.show()
```

**Review by tkoyama010 — APPROVED** [review]
LGTM with small suggestion. Thanks for adding this.

File: examples/01-filter/surface_reconstruction.py (hunk 2 above)
Comment (akaszynski) [reply examples/01-filter/surface_reconstruction.py:30]: This needs #1618.  I'll commit this suggestion once that's merged.

File: examples/01-filter/surface_reconstruction.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+"""
+.. _surface_reconstruction_example:
+
+Surface Reconstruction
+~~~~~~~~~~~~~~~~~~~~~~
+
+Surface reconstruction has a dedicated filter in PyVista and is
+handled by :func:`pyvista.DataSetFilters.reconstruct_surface`.  This
+tends to perform much better than :func:`DataSetFilters.delaunay_3d`.
```
Comment (MatthewFlamm) [inline examples/01-filter/surface_reconstruction.py:9]: Should we link to this new filter from `DataSetFilters.delaunay_3d`?

Comment (akaszynski) [reply examples/01-filter/surface_reconstruction.py:9]: Yes, that's a good point.

**Comment by akaszynski** [comment]
For record, also ran through the links with the sphinx linkcheck and fixed several of them.

**Pull request merged**